### PR TITLE
residuals() and sdf_residuals() for lm/glm

### PR DESCRIPTION
**Pull request by kevinykuo**
This PR implements `residuals()` and `sdf_residual()` methods for Spark ML linear regression and GLM regression models. The former returns a R vector and the latter returns a `tbl_spark`.

After looking at the MLlib code I discovered that it wasn't possible to simply "select" a residuals column *in addition* to the modeling DataFrame. The methods to calculate the ingredients for residuals (e.g. deviance) were not exposed in the API so we can't invoke them directly (or via reflection.) Instead of implementing a family/distribution hierarchy in sparklyr, I'm proposing that we "cbind" the output that's provided by the Spark API (i.e. a DataFrame with one column) with the the modeling dataset. Since the residuals output from Spark and the training dataset have the same distribution/order in the cluster (it's just a udf map), we can invoke `rdd.zip` which doesn't involve shuffling. This mechanism is implemented in the unexported function `sdf_fast_bind_cols()`.

The naming of the output column `"residuals"` was chosen to be consistent with `"predictions"` in `sdf_predict()`.

Related: https://github.com/rstudio/sparklyr/issues/670, https://github.com/rstudio/sparklyr/issues/557

Previous PR: https://github.com/rstudio/sparklyr/pull/672

### CODE REVIEW

File: R/sdf_utils.R [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#' This is a version of `sdf_bind_cols` that works by zipping
+#' RDDs. From the API docs: "Assumes that the two RDDs have the
+#' *same number of partitions* and the *same number of elements
+#' in each partition* (e.g. one was made through a map on the
+#' other)."
+#'
+#' @keywords internal
+#'
+#' @param ... Spark DataFrames to cbind
+sdf_fast_bind_cols <- function(...) {
```
Comment (kevinushey) [inline R/sdf_utils.R]: Is it possible for us to validate the requirement above re: two RDDs having same number of partitions + same number of elements in each partition? (Just to future-proof us in case we accidentally use this function where the above doesn't hold)

File: R/sdf_utils.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    spark_dataframe() %>%
+    spark_connection()
+
+  zip_sdf <- function(df1, df2) invoke_static(sc,
+                           "sparklyr.Utils",
+                           "zipDataFrames",
+                           spark_context(sc),
+                           df1, df2)
+
+  Reduce(zip_sdf, lapply(dots, spark_dataframe)) %>%
```
Comment (kevinushey) [inline R/sdf_utils.R]: Nice use of `Reduce` here :)

File: R/sdf_utils.R (hunk 1 above)
Comment (kevinykuo) [inline R/sdf_utils.R]: Checking the number of partitions is easy. Counting elements in each partition is more expensive since we have to iterate through them (albeit in parallel.) Maybe we can have a `validate` argument in the function that defaults to `TRUE`?

Comment (kevinushey) [inline R/sdf_utils.R]: I think that sounds good. Alternatively, we could hide the validation code behind some R / sparklyr configuration option?

Either way, maybe let's leave that for a separate PR in case we find that we want to use `sdf_fast_bind_cols()` in other contexts where it's less obvious that the required preconditions are met (for some reason).
